staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go

```diff
@@ -256,3 +240,138 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 		},
 	}, true, nil
 })
+
+func newCertificateVerifier(caProvider dynamiccertificates.CAContentProvider) *certificateVerifier {
+	out := &certificateVerifier{
+		caProvider: caProvider,
+		// We start at 1 since it's not the zero value
+		// cachedVerification.generation.
+		generation: 1,
+	}
+	caProvider.AddListener(out)
+	out.ctxCacheKey = cachedVerificationCtxKey(unsafe.Pointer(out))
+	return out
+}
+
+type certificateVerifier struct {
+	caProvider  dynamiccertificates.CAContentProvider
+	generation  uint64
+	ctxCacheKey cachedVerificationCtxKey
+}
+
+func (cv *certificateVerifier) verifyClientCerts(ctx context.Context, opts x509.VerifyOptions, peerCertificates []*x509.Certificate) ([][]*x509.Certificate, error) {
+	verify := func() ([][]*x509.Certificate, error) {
+		// Use intermediates, if provided
+		if opts.Intermediates == nil && len(peerCertificates) > 1 {
+			opts.Intermediates = x509.NewCertPool()
+			for _, intermediate := range peerCertificates[1:] {
+				opts.Intermediates.AddCert(intermediate)
+			}
+		}
+
+		return peerCertificates[0].Verify(opts)
+	}
+
+	entry, ok := ctx.Value(cv.ctxCacheKey).(*cachedVerification)
+	if !ok {
+		return verify()
+	}
+
+	if opts.CurrentTime.IsZero() {
+		opts.CurrentTime = time.Now()
+	}
+	currentGeneration := atomic.LoadUint64(&cv.generation)
+
+	entry.Lock()
```

Comment:
Oh this is incredible! I suppose this is even better than the striped cache approach you recommended because each entry itself is non-blocking from all other entries 😃 really great implementation @mikedanese, let me know if there's anything I can do to help move this proposal along, this is a huge improvement of the patch I'm carrying internally.